pkg/unidling/controller/controller.go

```diff
@@ -0,0 +1,323 @@
+		return true
+	}
+
+	// check to see if we think the error was transient (e.g. server error on the update request),
+	// and if not, do not retry
+	if !retry {
+		utilruntime.HandleError(fmt.Errorf("Unable to process unidling event for %s/%s at (%s), will not retry: %v", info.Namespace, info.Name, lastFired, err))
+		return true
+	}
+
+	// Otherwise, if we have an error, we were at least partially unsucessful in unidling, so
+	// we requeue the event to process later
+
+	// don't try to process failing requests forever
+	if c.queue.NumRequeues(infoRaw) > MaxRetries {
+		utilruntime.HandleError(fmt.Errorf("Unable to process unidling event for %s/%s (at %s), will not retry again: %v", info.Namespace, info.Name, lastFired, err))
+		c.queue.Forget(infoRaw)
+		return true
+	}
+
+	glog.V(4).Infof("Unable to fully process unidling request for %s/%s (at %s), will retry: %v", info.Namespace, info.Name, lastFired, err)
+	c.queue.AddRateLimited(infoRaw)
+	return true
+}
+
+// handleRequest handles a single request to unidle.  After checking the validity of the request,
+// it will examine the endpoints in question to determine which scalables to scale, and will scale
+// them and remove them from the endpoints' list of idled scalables.  If it is unable to properly
+// process the request, it will return a boolean indicating whether or not we should retry later,
+// as well as an error (e.g. if we're unable to parse an annotation, retrying later won't help,
+// so it will return false).
+func (c *UnidlingController) handleRequest(info types.NamespacedName, lastFired time.Time) (bool, error) {
+	// fetch the endpoints associated with the service in question
+	targetEndpoints, err := c.endpointsNamespacer.Endpoints(info.Namespace).Get(info.Name)
+	if err != nil {
+		return true, fmt.Errorf("unable to retrieve endpoints: %v", err)
+	}
+
+	// make sure we actually were idled...
+	idledTimeRaw, wasIdled := targetEndpoints.Annotations[unidlingapi.IdledAtAnnotation]
+	if !wasIdled {
+		glog.V(5).Infof("UnidlingController received a NeedPods event for a service that was not idled, ignoring")
+		return false, nil
+	}
+
+	// ...and make sure this request was to wake up from the most recent idling, and not a previous one
+	idledTime, err := time.Parse(time.RFC3339, idledTimeRaw)
+	if err != nil {
+		// retrying here won't help, we're just stuck as idle since we can't get parse the idled time
+		return false, fmt.Errorf("unable to check idled-at time: %v", err)
+	}
+	if lastFired.Before(idledTime) {
+		glog.V(5).Infof("UnidlingController received an out-of-date NeedPods event, ignoring")
+		return false, nil
+	}
+
+	// TODO: ew, this is unversioned.  Such is life when working with annotations.
+	var targetScalables []unidlingapi.RecordedScaleReference
+	if targetScalablesStr, hasTargetScalables := targetEndpoints.Annotations[unidlingapi.UnidleTargetAnnotation]; hasTargetScalables {
+		if err = json.Unmarshal([]byte(targetScalablesStr), &targetScalables); err != nil {
```

Comment:
You shouldn't be able to unmarshal from an internal object because you shouldn't use json tags in internal objects. Why can't you serialize the versioned representation of the object into the annotation?